atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java

```diff
@@ -451,26 +465,82 @@ private static LockAndTimestampServices createRawLeaderServices(
+        TimestampService localTime = AwaitingLeadershipProxy.newProxyInstance(TimestampService.class, time, leader);
+        env.register(localLock);
+        env.register(localTime);
+
+        // Create remote services, that may end up calling our own local services.
         Optional<SSLSocketFactory> sslSocketFactory = ServiceCreator.createSslSocketFactory(
                 leaderConfig.sslConfiguration());
+        RemoteLockService remoteLock = ServiceCreator.createService(
+                sslSocketFactory,
+                leaderConfig.leaders(),
+                RemoteLockService.class,
+                userAgent);
+        TimestampService remoteTime = ServiceCreator.createService(
+                sslSocketFactory,
+                leaderConfig.leaders(),
+                TimestampService.class,
+                userAgent);
 
-        return ImmutableLockAndTimestampServices.builder()
-                .lock(ServiceCreator.createService(
-                        sslSocketFactory,
-                        leaderConfig.leaders(),
-                        RemoteLockService.class,
-                        userAgent))
-                .time(ServiceCreator.createService(
-                        sslSocketFactory,
-                        leaderConfig.leaders(),
-                        TimestampService.class,
-                        userAgent))
-                .build();
+        if (leaderConfig.leaders().size() == 1) {
+            // Attempting to connect to ourself while processing a request can lead to deadlock if incoming request
+            // volume is high, as all Jetty threads end up waiting for the timestamp server, and no threads remain to
+            // actually handle the timestamp server requests. If we are the only single leader, we can avoid the
+            // deadlock entirely; so use PingableLeader's getUUID() to detect this situation and eliminate the redundant
+            // call.
+
+            PingableLeader localPingableLeader = localPaxosServices.pingableLeader();
+            String localServerId = localPingableLeader.getUUID();
+            PingableLeader remotePingableLeader = AtlasDbFeignTargetFactory.createRsProxy(
+                    sslSocketFactory,
+                    Iterables.getOnlyElement(leaderConfig.leaders()),
+                    PingableLeader.class,
+                    userAgent);
+
+            // Determine asynchronously whether the remote services are talking to our local services.
+            CompletableFuture<Boolean> useLocalServicesFuture = new CompletableFuture<>();
+            runAsync.accept(() -> {
+                int logAfter = LOGGING_INTERVAL;
+                while (true) {
+                    try {
+                        String remoteServerId = remotePingableLeader.getUUID();
+                        useLocalServicesFuture.complete(localServerId.equals(remoteServerId));
+                        return;
+                    } catch (ClientErrorException e) {
```

Comment:
does this indicate a 400 response? i would almost rather just do something dumb and obvious here like: retry for up to 1 minute to contact the remote - if no successful call can be made then give up.

realistically, if this is the local server it should succeed fairly quickly